SDN-5473: OVN-K: Port e2e tests for CUDN functionality, connectivity and isolation

ormergi opened the pull request:
This PR adds e2e tests for OVN-Kuberentes CluserUserDefinedNetwork CRD, introduced by https://github.com/ovn-kubernetes/ovn-kubernetes/pull/4612.

Note to reviewer:
- Tests has been added as is from U/S repo, no logical changes has been done (1st & 2nd commits)
  Any change required on those bits, require introducing the change on U/S and replicating these changes to D/S, after this PR is merged. In order to avoid diverging from U/S code.
- Due to flakes surfaced on OCP CI the some changes has been done on U/S https://github.com/ovn-kubernetes/ovn-kubernetes/pull/4842, this PR changes were replicated to this PR as is (3-5 commits).
Any change required on those bits, require introducing the change on the PR on U/S and replicating these changes to this PR. In order to avoid diverging from U/S code.
The changes has been done:
      - Add missing wait for CUDN status to populate.
      - Since tests runs in parallel on TP lane, wait for CUDN objects to be gone when deleted.
      - Prevent flakes when CUDN realted tests run in parallel by using random metadata.name for CUDN objects.
- In tests that create namespaces and pods, set the pods with PSA that meet the cluster.
- Following incoming changes related to [condition renaming ](https://github.com/ovn-kubernetes/ovn-kubernetes/pull/4884) the new CUDN tests were adjusted (last commit), similar to what we did at https://github.com/openshift/origin/pull/29381
  
- ~~This PR depends on having the CRD deployed by CNO https://github.com/openshift/cluster-network-operator/pull/2558~~

ormergi commented:
/test ?

ormergi commented:
/test e2e-gcp-ovn-techpreview

jluhrsen commented:
your  tests are not registered with the label `[Suite:openshift/conformance/parallel]` and that's what we are running in that job with `openshift-tests run openshift/conformance/parallel`.
You can get that fixed by running `make update`
you can check locally with this command:

before make update
```
❯ ./openshift-tests run openshift/conformance/parallel --dry-run | rg 'should create NAD according to spec in each target namespace and report active namespaces'
I1114 14:15:55.181022 1641198 factory.go:193] Registered Plugin "containerd"
INFO[0000] Decoding provider                             clusterState="<nil>" discover=true dryRun=true func=DecodeProvider providerType=
```
after make update
```
❯ ./openshift-tests run openshift/conformance/parallel --dry-run | rg 'should create NAD according to spec in each target namespace and report active namespaces'
I1114 14:18:09.348749 1642240 factory.go:193] Registered Plugin "containerd"
INFO[0000] Decoding provider                             clusterState="<nil>" discover=true dryRun=true func=DecodeProvider providerType=
"[sig-network][OCPFeatureGate:NetworkSegmentation][Feature:UserDefinedPrimaryNetworks] when using openshift ovn-kubernetes ClusterUserDefinedNetwork CRD Controller should create NAD according to spec in each target namespace and report active namespaces [Suite:openshift/conformance/parallel]" 
```

ormergi commented:
@jluhrsen thanks you!

ormergi commented:
/test e2e-gcp-ovn-techpreview

ormergi commented:
/test e2e-gcp-ovn-techpreview

ormergi commented:
/test e2e-gcp-ovn-techpreview

ormergi commented:
/test e2e-gcp-ovn-techpreview

ormergi commented:
/test e2e-gcp-ovn-techpreview

ormergi commented:
/test e2e-gcp-ovn-techpreview

ormergi commented:
/test e2e-gcp-ovn-techpreview

ormergi commented:
https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/29274/pull-ci-openshift-origin-master-e2e-gcp-ovn-techpreview/1861012229453778944

Failures seem not related to this PR changes

maiqueb commented on test/extended/networking/network_segmentation.go:
```diff
@@ -125,7 +124,7 @@ var _ = Describe("[sig-network][OCPFeatureGate:NetworkSegmentation][Feature:User
 						"for two pods connected over a L2 primary UDN",
 						networkAttachmentConfigParams{
-							name:     nadName,
+							name:     randomNetworkMetaName(),
```
there's no point doing this for NADs, since they already randomize the `networkName` attribute - which is the ID of the network.


for UDNs (cluster or not) I think you *do* need to randomize.

ormergi replied:
Fixed, following discussion at https://github.com/ovn-kubernetes/ovn-kubernetes/pull/4842#discussion_r1866845321

maiqueb reviewed (changes requested):
Thank you.

ormergi commented on test/extended/networking/network_segmentation.go:
```diff
@@ -682,6 +682,266 @@ var _ = Describe("[sig-network][OCPFeatureGate:NetworkSegmentation][Feature:User
+				cleanup, err := createManifest("", newClusterUDNManifest(testClusterUdnName, testTenantNamespaces...))
+				DeferCleanup(func() error {
+					cleanup()
+					_, _ = e2ekubectl.RunKubectl("", "delete", clusterUserDefinedNetworkResource, testClusterUdnName)
```
It may return not found error which I think should go silent. 
In follow-up commit there is additional logic to ensure the object is gone.
Make sense?

maiqueb replied:
Hmpf .... I don't like it. But... OK.

Let's at least be honest here and change the return type of the fn `DeferCleanup` invokes, ok ? I.e. no need to return an error since you don't do anything with it.

ormergi replied:
Done

ormergi replied:
Fixed in follow up commit

maiqueb replied:
I still see it ... You're still returning the error here, only to have a single `return nil` clause below, in line 745

ormergi commented on test/extended/networking/network_segmentation.go:
```diff
@@ -877,6 +1142,117 @@ func normalizeConditions(conditions []metav1.Condition) []metav1.Condition {
+
+	ExpectWithOffset(1, nad.Name).To(Equal(udnName))
+	ExpectWithOffset(1, nad.Namespace).To(Equal(namespace))
+	ExpectWithOffset(1, nad.OwnerReferences).To(Equal([]metav1.OwnerReference{{
```
Initially we used Equal because we expect the CUDN to have single owner-ref.

maiqueb replied:
So ? You can use `ConsistOf` and pass a single owner ref. 

It's the same, but avoids have to specify an array of owner references.

Not a must, but the code looks better.

ormergi commented on test/extended/networking/network_segmentation.go:
```diff
@@ -945,13 +959,14 @@ var _ = Describe("[sig-network][OCPFeatureGate:NetworkSegmentation][Feature:User
 	})
 })
 
+var nadToUdnParams = map[string]string{
+	"primary":   "Primary",
+	"secondary": "Secondary",
+	"layer2":    "Layer2",
+	"layer3":    "Layer3",
+}
```
Actually to the format we use in UDN/CUDN API (pascal case).
They come to have the same names, which are single word so it seems capitulated :nerd_face:

maiqueb replied:
Asking in case it made sense to just pass a func to map to all values that would capitalize the first character. 

Not sure it makes sense - keep what you have.

ormergi commented on test/extended/networking/network_segmentation.go:
```diff
@@ -558,8 +558,11 @@ var _ = Describe("[sig-network][OCPFeatureGate:NetworkSegmentation][Feature:User
+					Eventually(func() bool {
+						_, err := e2ekubectl.RunKubectl("", "get", "clusteruserdefinednetwork", c.name)
+						return strings.Contains(err.Error(), fmt.Sprintf("clusteruserdefinednetworks.k8s.ovn.org %q not found", c.name))
+					}, 1*time.Minute, 3*time.Second).Should(BeTrueBecause("ClusterUserDefinedNetwork %q should be gone", c.name))
```
Fixed

ormergi commented on test/extended/networking/network_segmentation.go:
```diff
@@ -877,6 +1142,117 @@ func normalizeConditions(conditions []metav1.Condition) []metav1.Condition {
+	conditions = normalizeConditions(conditions)
+	expectedMsg := fmt.Sprintf("failed to delete NetworkAttachmentDefinition [%[1]s/%[2]s]: network in use by the following pods: [%[1]s/%[3]s]",
+		udnNamespace, udnName, expectedPodName)
+	ExpectWithOffset(1, conditions).To(Equal([]metav1.Condition{
```
Initially we used Equal because in those test we expect single condition

ormergi commented on test/extended/networking/network_segmentation.go:
```diff
@@ -550,6 +550,20 @@ var _ = Describe("[sig-network][OCPFeatureGate:NetworkSegmentation][Feature:User
+				cudnManifest := generateClusterUserDefinedNetworkManifest(&c)
+				cleanup, err := createManifest("", cudnManifest)
+				DeferCleanup(func() {
+					cleanup()
```
According to ginkgo docs it should be used in tests helpers.
We use DeferCleanup in other projects as well but we didnt need the GinkgoHelper.
I recall that if the DeferCleanup blows up we get proper stack trace with meaningful code references.

maiqueb replied:
> I recall that if the DeferCleanup blows up we get proper stack trace with meaningful code references.

Can you provide a link for that ? Never heard of it.

ormergi replied:
Its not documented anywhere, it acts like a BeforeEach, for example: here is what we see on Ginkgo output when   DeferCleanup fail 
```
  [FAILED] in [DeferCleanup (Each)] - /home/omergi/workspace/github.com/ovn-kubernetes/test/e2e/network_segmentation.go:1057 @ 12/10/24 18:03:31.354
```

maiqueb commented on test/extended/networking/network_segmentation.go:
```diff
@@ -682,16 +706,295 @@ var _ = Describe("[sig-network][OCPFeatureGate:NetworkSegmentation][Feature:User
+					Eventually(func() bool {
+						_, err := e2ekubectl.RunKubectl("", "get", clusterUserDefinedNetworkResource, testClusterUdnName)
+						return strings.Contains(err.Error(), fmt.Sprintf("clusteruserdefinednetworks.k8s.ovn.org %q not found", testClusterUdnName))
+					}, 1*time.Minute, 3*time.Second).Should(BeTrueBecause("ClusterUserDefinedNetwork %q should be gone", testClusterUdnName))
```
shouldn't we match error instead of this ? 

It's so much more readable ...

ormergi replied:
I wish we could, the `e2ekubectl.RunKubectl` doesnt play nice with the error matcher, 
because `e2ekubectl.RunKubectl` return some custom error in a formatting that makes it impossible to convert to some standard k8s error.

maiqueb replied:
OK, I had just glanced, and missed the fact there's actually a reason to treat the error as a string ...

maiqueb commented on test/extended/networking/network_segmentation.go:
```diff
@@ -682,6 +682,266 @@ var _ = Describe("[sig-network][OCPFeatureGate:NetworkSegmentation][Feature:User
+					Eventually(func() bool {
+						_, err := nadClient.NetworkAttachmentDefinitions(nsName).Get(context.Background(), testClusterUdnName, metav1.GetOptions{})
+						return err != nil && kerrors.IsNotFound(err)
+					}, time.Second*3, time.Second*1).Should(BeTrue(),
```
nit: the `*1` in `time.Second*1` is redundant ...

maiqueb commented on test/extended/networking/network_segmentation.go:
```diff
@@ -877,6 +1142,117 @@ func normalizeConditions(conditions []metav1.Condition) []metav1.Condition {
 }
 
+func assertClusterNADManifest(nadClient nadclient.K8sCniCncfIoV1Interface, namespace, udnName, udnUID string) {
+	nad, err := nadClient.NetworkAttachmentDefinitions(namespace).Get(context.Background(), udnName, metav1.GetOptions{})
```
please use `GinkgoHelper`. This way you don't need to be counting offsets.

ormergi commented:
[Changes](https://github.com/openshift/origin/compare/54d3a296d1ade5ed2e8952dde8e412f7ff459a7b..2c6a26644d994eb1b9423bf62d16edf8a120ec0a): resolve https://github.com/openshift/origin/pull/29274#discussion_r1888867716

ormergi commented:
@maiqueb thanks for the review :slightly_smiling_face: 
Regarding the comments on the first two commits (adds the new e2e tests), as mentioned in the PR description, these commits were taken from U/S as is - in order to avoid diverging from U/S code.

Running those tests on OCP CI surfaced flakes we didn't saw on U/S CI, requiring some additional changes to stabilize the new tests  (in the other commits 3-6).
In order avoid diverging from U/S code, I posted a PR on U/S with the additional changes https://github.com/ovn-kubernetes/ovn-kubernetes/pull/4842 and replicated these changes to this PR (thanks for the review on that PR as well).

I hope we can defer addressing the comments of the first two commit to follow-up PR, I think these changes should go  "through U/S", to avoid diverging for U/S.
As for the rest of commits, I addressed the comments you posted here on the U/S PR and replicated that changes to this PR.

maiqueb commented on test/extended/networking/network_segmentation.go:
```diff
@@ -682,16 +706,294 @@ var _ = Describe("[sig-network][OCPFeatureGate:NetworkSegmentation][Feature:User
+				patch := fmt.Sprintf(`[{"op": "add", "path": "./spec/namespaceSelector/matchExpressions/0/values/-", "value": "%s"}]`, testNewNs)
+				_, err := e2ekubectl.RunKubectl("", "patch", clusterUserDefinedNetworkResource, testClusterUdnName, "--type=json", "-p="+patch)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(waitForClusterUserDefinedNetworkReady(testClusterUdnName, 5*time.Second)).To(Succeed())
```
from what I'm seeing here, it seems that if a network has namespaces referenced in its selector that don't exist that it'll still reach the `ready` condition. Is this intentional ? 

i.e. please remind me .... what's the condition for a network to be ready ? ... I assumed it would be for it to be available on *all* the namespaces it lists in its namespace selector.

ormergi replied:
> from what I'm seeing here, it seems that if a network has namespaces referenced in its selector that don't exist that it'll still reach the ready condition. Is this intentional ?

Yes, by design.
Once a new namespace is created or existing one is selected (i.e.: has a label that applies to the selector) and applies to the selector, the controller will create NAD in the new namespace.
In case a selected namespace is deleted or selection removed (i.e.: label that applies to the selector is removed), the controller will delete the NAD in this namespace.

> i.e. please remind me .... what's the condition for a network to be ready ? ... I assumed it would be for it to be available on all the namespaces it lists in its namespace selector.

Yes this is basically what it means.
Once all NAD exist in all namespaces that applies for the selector ready is true.
In case no namespace match the selector, the condition is set to true.

maiqueb commented on test/extended/networking/network_segmentation.go:
```diff
@@ -682,16 +706,294 @@ var _ = Describe("[sig-network][OCPFeatureGate:NetworkSegmentation][Feature:User
+					Eventually(func() bool {
+						_, err := nadClient.NetworkAttachmentDefinitions(removedTenantNs).Get(context.Background(), testClusterUdnName, metav1.GetOptions{})
+						return err != nil && kerrors.IsNotFound(err)
+					}, time.Second*300, time.Second*1).Should(BeTrue(),
```
do we really need to check every second ? really asking. If we're giving it 5 minutes to fail, we could maybe get away w/ asking the API only now and then. Maybe 5 / 10 secs period ?

maiqueb commented on test/extended/networking/network_segmentation.go:
```diff
@@ -682,16 +706,294 @@ var _ = Describe("[sig-network][OCPFeatureGate:NetworkSegmentation][Feature:User
+					cmd := e2ekubectl.NewKubectlCommand("", "delete", clusterUserDefinedNetworkResource, testClusterUdnName)
+					cmd.WithTimeout(time.NewTimer(deleteNetworkTimeout).C)
+					_, err := cmd.Exec()
+					Expect(err).To(HaveOccurred(), "should fail to delete ClusterUserDefinedNetwork when used")
```
nit: could we match the error ?

maiqueb commented on test/extended/networking/network_segmentation.go:
```diff
@@ -682,16 +706,294 @@ var _ = Describe("[sig-network][OCPFeatureGate:NetworkSegmentation][Feature:User
+						_ = nadClient.NetworkAttachmentDefinitions(inUseNetTestTenantNamespace).Delete(ctx, testClusterUdnName, metav1.DeleteOptions{})
+						_, err := nadClient.NetworkAttachmentDefinitions(inUseNetTestTenantNamespace).Get(ctx, testClusterUdnName, metav1.GetOptions{})
+						return err
+					}).ShouldNot(HaveOccurred(),
```
nit: ah. Shouldn't we assert the NAD data instead ?

I'm asking because it took me a while to figure out what you were doing. Not needed, maybe I'm just a bit thick today.

maiqueb commented on test/extended/networking/network_segmentation.go:
```diff
@@ -682,16 +706,294 @@ var _ = Describe("[sig-network][OCPFeatureGate:NetworkSegmentation][Feature:User
+					Eventually(func() error {
+						_, err := e2ekubectl.RunKubectl("", "get", clusterUserDefinedNetworkResource, testClusterUdnName)
+						return err
+					}, udnInUseDeleteTimeout, deleteNetworkInterval).Should(HaveOccurred(),
```
nit: maybe match the error

jluhrsen commented:
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-aws-ovn-hypershift-conformance-techpreview
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-azure-ovn-techpreview
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-dualstack-techpreview
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-vsphere-ovn-techpreview
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-aws-ovn-techpreview
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-gcp-ovn-techpreview
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-techpreview

maiqueb reviewed (commented):
I've left a ton of nits, but I need you to confirm these 2 for me:
- https://github.com/openshift/origin/pull/29274#discussion_r1891878937
- https://github.com/openshift/origin/pull/29274#discussion_r1891881221

Depending on this I can lgtm.

ormergi commented on test/extended/networking/network_segmentation.go:
```diff
@@ -682,16 +706,294 @@ var _ = Describe("[sig-network][OCPFeatureGate:NetworkSegmentation][Feature:User
+				testUdnUID := strings.Trim(udnUidRaw, "'")
+
+				By("verify a NAD exist in new namespace according to spec")
+				assertClusterNADManifest(nadClient, testNewNs, testClusterUdnName, testUdnUID)
```
The network ready condition indicated the controller create NAD in namespaces that applies to the selector.

ormergi commented:
Rebased

ormergi commented:
[Changes](https://github.com/openshift/origin/compare/2381d0e25ff0ce3b08de67f53aea9dbefa889aeb..93043d76b3afdd9a919706745d73389688730fa0): bring latest changes of https://github.com/ovn-kubernetes/ovn-kubernetes/pull/4842, following CI failure fixes.

jluhrsen commented:
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-aws-ovn-techpreview
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-azure-ovn-techpreview
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-gcp-ovn-techpreview-upgrade
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-techpreview
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-ipv6-techpreview
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-dualstack-techpreview

jluhrsen commented:
the unit tests have failure(s)

/test unit

tssurya commented:
/test e2e-gcp-ovn-techpreview

tssurya commented:
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-ipv6-techpreview

ormergi commented:
[Rebase](https://github.com/openshift/origin/compare/93043d76b3afdd9a919706745d73389688730fa0..80ab01547d798fc1b544b1dd1c31dcaed708450c)

ormergi commented:
Pushed new commit to adjust the introduced CUDN CRD tests to fit incomming condition type name by [1], similar to [2].

[1] https://github.com/ovn-kubernetes/ovn-kubernetes/pull/4884
[2] https://github.com/openshift/origin/pull/29381

ormergi commented:
/unhold

ormergi commented:
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-aws-ovn-techpreview
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-azure-ovn-techpreview
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-gcp-ovn-techpreview-upgrade
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-techpreview
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-ipv6-techpreview
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-dualstack-techpreview

tssurya commented:
/test e2e-gcp-ovn-techpreview

ormergi commented:
@tssurya @jluhrsen  the[ payload jobs passed](https://github.com/openshift/origin/pull/29274#issuecomment-2579651603), except for `periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-ipv6-techpreview`. It seem to failed due to unrelated responses to this PR changes.

tssurya commented:
/label acknowledge-critical-fixes-only

jluhrsen commented:
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-techpreview
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-ipv6-techpreview
/payload-job periodic-ci-openshift-release-master-nightly-4.19-e2e-metal-ipi-ovn-dualstack-techpreview

jluhrsen commented:
/test e2e-gcp-ovn-techpreview-upgrade

jluhrsen commented:
/payload-job periodic-ci-openshift-release-master-ci-4.18-e2e-aws-ovn-techpreview
/payload-job periodic-ci-openshift-release-master-ci-4.18-e2e-aws-ovn-techpreview-serial

maiqueb commented on test/extended/networking/network_segmentation.go:
```diff
@@ -686,6 +707,281 @@ var _ = Describe("[sig-network][OCPFeatureGate:NetworkSegmentation][Feature:User
+					_, err := cs.CoreV1().Namespaces().Create(context.Background(), &v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: nsName}}, metav1.CreateOptions{})
+					Expect(err).NotTo(HaveOccurred())
+					DeferCleanup(func() error {
+						err := cs.CoreV1().Namespaces().Delete(context.Background(), nsName, metav1.DeleteOptions{})
+						return err
```
nit

```suggestion
						return cs.CoreV1().Namespaces().Delete(context.Background(), nsName, metav1.DeleteOptions{})
```

maiqueb reviewed (approved):
left two more nits, but I think we can live with them ...

ormergi commented:
> left two more nits, but I think we can live with them ...

@maiqueb thanks for the review I prefer to fix these nits and bunch of other comments in follow up commit (through U/S).
If you OK with that, please approve :slightly_smiling_face:

maiqueb commented:
> > left two more nits, but I think we can live with them ...
> 
> @maiqueb thanks for the review I prefer to fix these nits and bunch of other comments in follow up PR (through U/S). If you OK with that, please approve 🙂

Didn't I already ?... I'm a nobody in this repo, hence you won't get tags from me unfortunately ...

/lgtm

dgoodwin commented:
/approve

tssurya commented:
/test e2e-gcp-ovn

tssurya commented:
/tide refresh

tssurya commented:
/test e2e-aws-ovn-serial]

tssurya commented:
/test e2e-aws-ovn-serial

tssurya commented:
/test e2e-aws-ovn-serial

tssurya commented:
/tide refresh

tssurya commented:
/tide refresh

jluhrsen commented:
/test e2e-aws-ovn-serial

ormergi commented:
/test e2e-aws-ovn-serial

ormergi commented:
/tide refresh

ormergi commented:
/test e2e-aws-ovn-serial

dgoodwin commented:
/override ci/prow/e2e-aws-ovn-serial

tssurya commented:
/tide refresh

tssurya commented:
/test e2e-gcp-ovn-upgrade

dgoodwin commented:
/override ci/prow/e2e-gcp-ovn-upgrade

jluhrsen commented:
/tide refresh

stbenjam commented:
/skip

Just making sure all the failed ones are optional

stbenjam commented:
/override ci/prow/e2e-aws-ovn-serial

tssurya commented:
/tide refresh

jluhrsen commented:
/test e2e-metal-ipi-ovn-ipv6

jluhrsen commented:
/test e2e-aws-ovn-serial
